File: 438.py
from collections import Counter

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        letters = Counter()
        target = Counter(p)
        left = 0
        res = []

        for index, letter in enumerate(s):
            letters[letter] += 1
            if(len(p) == index - left + 1):
                if(letters == target):
                    res.append(left)
                letters[s[left]] -= 1
                if(letters[s[left]] == 0):
                    del letters[s[left]]
                left += 1

        return res

File: 438.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        unordered_map<char, int> count_s, count_p;
        int matches = 0, left = 0, count;
        vector<int> res;

        for(int i = 0; i < p.size(); i++) {
            count_p[p[i]]++;
        }
        count = count_p.size();

        for(int i = 0; i < s.size(); i++) {
            count_s[s[i]]++;
            if(count_s[s[i]] == count_p[s[i]]) {
                matches++;
            }
            if(i - left + 1 == p.size()) {
                if(matches == count) {
                    res.push_back(left);
                }
                if(count_s[s[left]] == count_p[s[left]]) {
                    matches--;
                }
                count_s[s[left]]--;
                if(count_s[s[left]] == 0) {
                    count_s.erase(s[left]);
                }
                left++;
            }
        }
        return res;
    }
};